### DISCUSSION TITLE
Performance / JVM restart / dbus (Windows)?

### DISCUSSION THREAD

**Discussion by JuhaFin69** [AsamK/signal-cli#471] [Q&A] [answered]
Hi! I am using signal-cli in Windows. In my application performance does matter, which for my application means how much time sending of multiple messages(10-20) takes. I do not know much of Java. I assume that every signal-cli.bat call will restart JVM again and therefore signal-cli calls take some time(2-3 seconds). Is there any option to configure Java so that shutdown of JVM will not happen and all calls would be made by same JVM instance? Or should I install this dbus [https://gitlab.freedesktop.org/dbus/dbus/-/tree/dbus-1.12](https://gitlab.freedesktop.org/dbus/dbus/-/tree/dbus-1.12) to my Windows machine and then use signal-cli daemon mode? Does this combo Windows & dbus & daemon mode work?

**Comment by exquo** [comment, marked as answer]
> I assume that every signal-cli.bat call will restart JVM again and therefore signal-cli calls take some time(2-3 seconds).

Correct.

> Is there any option to configure Java so that shutdown of JVM will not happen and all calls would be made by same JVM instance? 

Run the daemon and communicate over DBus, like you suggest. Or, now there's also an experimental [native build with GraalVM](https://github.com/AsamK/signal-cli/blob/master/README.md#user-content-building-a-native-binary-with-graalvm-experimental).

> Or should I install this dbus https://gitlab.freedesktop.org/dbus/dbus/-/tree/dbus-1.12 to my Windows machine and then use signal-cli daemon mode? Does this combo Windows & dbus & daemon mode work?

I would like to know that too! :) According to their doc at least, windows is supported.

See also [Performance FAQ](https://github.com/AsamK/signal-cli/wiki/FAQ#user-content-performance).

**Comment by skull-squadron** [comment]
Help me understand GraalVM tech with an ignorant Q: 

1. Does the native build not require a JVM at runtime? (I'm assuming just at compile time)

2. If "yes" to 1., then I assume this means the better startup performance on most platforms, yes?

**Reply by exquo** [reply]
Yes, to both.
